src/benchmark.rs

```diff
@@ -0,0 +1,74 @@
+//! Benchmark features to implement different specific benchmarks.
+//!
+//! This is useful for creating different benchmark on time of execution or
+//! the number of instruction for example.
+
+use core::arch::asm;
+
+use crate::arch::{Arch, Architecture, Csr};
+
+pub struct Benchmark {
+    // Does the benchmark started.
+    pub is_running: bool,
+
+    // Temporary value to store previous state (e.g. state when the benchmark started to compare).
+    pub previous_instr_count: usize,
+
+    pub previous_timer: usize,
+}
+
+impl Benchmark {
+    #[allow(dead_code)]
+    pub fn new() -> Benchmark {
+        Benchmark {
+            is_running: false,
+            previous_instr_count: 0,
+            previous_timer: 0,
+        }
+    }
+
+    /// Start benchmarking.
+    #[allow(dead_code)]
+    pub fn start(&mut self) {
+        if self.is_running {
+            log::error!("already counting instr");
+        } else {
+            self.is_running = true;
+
+            // instruction
+            self.previous_instr_count = Arch::read_csr(Csr::Minstret);
+
+            // time
+            unsafe {
+                asm!(
+                    "rdtime {value}",
+                    value = out(reg) self.previous_timer
+                );
+            }
+        }
+    }
+
+    /// Stop benchmarking.
+    #[allow(dead_code)]
+    pub fn stop(&mut self, tag: &str) {
+        self.is_running = false;
+
+        // instructions
+        let stop_value = Arch::read_csr(Csr::Minstret);
+        Self::record_entry("instr", tag, stop_value - self.previous_instr_count);
+
+        // time
+        let stop_time: usize;
+        unsafe {
+            asm!(
+                "rdtime {value}", 
+                value = out(reg) stop_time);
+        }
+        Self::record_entry("time", tag, stop_time - self.previous_timer);
+    }
+
+    /// Print formated string with value of the entry and a tag for identification.
+    fn record_entry(bench: &str, tag: &str, entry: usize) {
+        log::debug!("|| {:>15} || {:25} || {}", bench, tag, entry);
```

Comment:
if that is easier we could also use a custom print function for benchmarks. That way this would ignore the log level (which is something orthogonal to the benchmarks, and we might even want to benchmark the impact of the log level in the future).

To do that you could create a macro (e.g. `benchmark_print`) that uses `core::format_args!` and calls `Plat::debug_print`. There is an example in the Tyche repo [here](https://github.com/epfl-dcsl/tyche-devel/blob/ec97e692fe5a97d0f0b21cc4818f085e2ca197aa/monitor/tyche/src/debug.rs#L21-L32). If you do that be sure to check if benchmarks are enabled in the macro, so that nothing gets printed at all when benchmarks are disables ^^